File: backend/src/services/flow_service.py
"""Service layer for document flow operations."""

from sqlalchemy.orm import Session
from typing import Optional, List, Tuple
from fastapi import HTTPException, status

from ..models.flow import DocumentFlow, flow_question_groups
from ..models.question import QuestionGroup
from ..schemas.flow import DocumentFlowCreate, DocumentFlowUpdate


class FlowService:
    """Service for document flow operations."""
    
    @staticmethod
    def create_flow(
        db: Session,
        flow_data: DocumentFlowCreate,
        user_id: int
    ) -> DocumentFlow:
        """
        Create a new document flow.
        
        Args:
            db: Database session
            flow_data: Flow creation data
            user_id: User ID creating the flow
            
        Returns:
            Created document flow
        """
        # Check if flow name already exists
        existing = db.query(DocumentFlow).filter(
            DocumentFlow.name == flow_data.name,
            DocumentFlow.is_active == True
        ).first()
        
        if existing:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Flow with name '{flow_data.name}' already exists"
            )
        
        # Validate starting group if provided
        if flow_data.starting_group_id:
            starting_group = db.query(QuestionGroup).filter(
                QuestionGroup.id == flow_data.starting_group_id,
                QuestionGroup.is_active == True
            ).first()
            
            if not starting_group:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Starting group not found"
                )
        
        # Create flow
        flow = DocumentFlow(
            name=flow_data.name,
            description=flow_data.description,
            starting_group_id=flow_data.starting_group_id,
            created_by=user_id,
            is_active=True
        )
        
        db.add(flow)
        db.commit()
        db.refresh(flow)
        
        # Associate question groups if provided
        if flow_data.question_group_ids:
            FlowService._associate_question_groups(
                db,
                flow.id,
                flow_data.question_group_ids
            )
        
        return flow
    
    @staticmethod
    def _associate_question_groups(
        db: Session,
        flow_id: int,
        group_ids: List[int]
    ):
        """Associate question groups with a flow."""
        for idx, group_id in enumerate(group_ids):
            # Verify group exists
            group = db.query(QuestionGroup).filter(
                QuestionGroup.id == group_id,
                QuestionGroup.is_active == True
            ).first()
            
            if not group:
                continue
            
            # Insert association
            db.execute(
                flow_question_groups.insert().values(
                    flow_id=flow_id,
                    question_group_id=group_id,
                    order_index=idx
                )
            )
        
        db.commit()
    
    @staticmethod
    def get_flow(db: Session, flow_id: int) -> Optional[DocumentFlow]:
        """
        Get flow by ID.
        
        Args:
            db: Database session
            flow_id: Flow ID
            
        Returns:
            Flow if found, None otherwise
        """
        return db.query(DocumentFlow).filter(
            DocumentFlow.id == flow_id,
            DocumentFlow.is_active == True
        ).first()
    
    @staticmethod
    def list_flows(
        db: Session,
        skip: int = 0,
        limit: int = 100,
        search: Optional[str] = None
    ) -> Tuple[List[DocumentFlow], int]:
        """
        List document flows.
        
        Args:
            db: Database session
            skip: Number of records to skip
            limit: Maximum number of records to return
            search: Optional search term
            
        Returns:
            Tuple of (flows list, total count)
        """
        query = db.query(DocumentFlow).filter(
            DocumentFlow.is_active == True
        )
        
        if search:
            query = query.filter(
                DocumentFlow.name.ilike(f"%{search}%")
            )
        
        total = query.count()
        flows = query.order_by(DocumentFlow.created_at.desc()).offset(skip).limit(limit).all()
        
        return flows, total
    
    @staticmethod
    def update_flow(
        db: Session,
        flow_id: int,
        flow_data: DocumentFlowUpdate
    ) -> Optional[DocumentFlow]:
        """
        Update a document flow.
        
        Args:
            db: Database session
            flow_id: Flow ID
            flow_data: Flow update data
            
        Returns:
            Updated flow if found, None otherwise
        """
        flow = FlowService.get_flow(db, flow_id)
        if not flow:
            return None
        
        # Update fields
        if flow_data.name is not None:
            # Check name uniqueness
            existing = db.query(DocumentFlow).filter(
                DocumentFlow.name == flow_data.name,
                DocumentFlow.id != flow_id,
                DocumentFlow.is_active == True
            ).first()
            
            if existing:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Flow with name '{flow_data.name}' already exists"
                )
            
            flow.name = flow_data.name
        
        if flow_data.description is not None:
            flow.description = flow_data.description
        
        if flow_data.starting_group_id is not None:
            flow.starting_group_id = flow_data.starting_group_id
        
        # Update question group associations if provided
        if flow_data.question_group_ids is not None:
            # Remove existing associations
            db.execute(
                flow_question_groups.delete().where(
                    flow_question_groups.c.flow_id == flow_id
                )
            )
            
            # Add new associations
            FlowService._associate_question_groups(
                db,
                flow_id,
                flow_data.question_group_ids
            )
        
        db.commit()
        db.refresh(flow)
        
        return flow
    
    @staticmethod
    def delete_flow(db: Session, flow_id: int) -> bool:
        """
        Soft delete a document flow.
        
        Args:
            db: Database session
            flow_id: Flow ID
            
        Returns:
            True if deleted, False if not found
        """
        flow = FlowService.get_flow(db, flow_id)
        if not flow:
            return False
        
        flow.is_active = False
        db.commit()
        
        return True
    
    @staticmethod
    def get_flow_with_groups(db: Session, flow_id: int) -> Optional[dict]:
        """
        Get flow with associated question groups.
        
        Args:
            db: Database session
            flow_id: Flow ID
            
        Returns:
            Flow data with question groups
        """
        flow = FlowService.get_flow(db, flow_id)
        if not flow:
            return None
        
        # Get associated question groups
        groups = db.query(QuestionGroup).join(
            flow_question_groups,
            QuestionGroup.id == flow_question_groups.c.question_group_id
        ).filter(
            flow_question_groups.c.flow_id == flow_id,
            QuestionGroup.is_active == True
        ).order_by(flow_question_groups.c.order_index).all()
        
        return {
            **flow.to_dict(),
            "question_groups": [g.to_dict() for g in groups]
        }

File: go-backend/internal/services/flow_service.go
package services

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/estate-docs/go-backend/internal/models"
	"gorm.io/gorm"
)

type FlowService struct {
	DB *gorm.DB
}

func NewFlowService(db *gorm.DB) *FlowService {
	return &FlowService{DB: db}
}

type FlowCreateInput struct {
	Name             string  `json:"name"`
	Description      *string `json:"description"`
	StartingGroupID  *int    `json:"starting_group_id"`
	QuestionGroupIDs []int   `json:"question_group_ids"`
}

type FlowUpdateInput struct {
	Name             *string          `json:"name"`
	Description      *string          `json:"description"`
	StartingGroupID  *int             `json:"starting_group_id"`
	QuestionGroupIDs []int            `json:"question_group_ids"`
	FlowLogic        *json.RawMessage `json:"flow_logic"`
	IsActive         *bool            `json:"is_active"`
}

func (s *FlowService) Create(input FlowCreateInput, userID int) (*models.DocumentFlow, error) {
	f := models.DocumentFlow{
		Name:            input.Name,
		Description:     input.Description,
		StartingGroupID: input.StartingGroupID,
		CreatedBy:       &userID,
		IsActive:        true,
		CreatedAt:       time.Now().UTC(),
		UpdatedAt:       time.Now().UTC(),
	}

	if err := s.DB.Create(&f).Error; err != nil {
		return nil, err
	}

	// Associate question groups
	if len(input.QuestionGroupIDs) > 0 {
		for i, gid := range input.QuestionGroupIDs {
			fqg := models.FlowQuestionGroup{
				FlowID:          f.ID,
				QuestionGroupID: gid,
				OrderIndex:      i,
			}
			s.DB.Create(&fqg)
		}
	}

	return &f, nil
}

func (s *FlowService) List(skip, limit int, search *string) ([]models.DocumentFlow, int64, error) {
	var flows []models.DocumentFlow
	var total int64

	q := s.DB.Model(&models.DocumentFlow{}).Where("is_active = true")
	if search != nil && *search != "" {
		q = q.Where("name ILIKE ?", "%"+*search+"%")
	}
	q.Count(&total)

	if err := q.Order("id DESC").Offset(skip).Limit(limit).Find(&flows).Error; err != nil {
		return nil, 0, err
	}
	return flows, total, nil
}

func (s *FlowService) GetByID(id int) (*models.DocumentFlow, error) {
	var f models.DocumentFlow
	if err := s.DB.Where("is_active = true").First(&f, id).Error; err != nil {
		return nil, err
	}
	return &f, nil
}

func (s *FlowService) GetWithGroups(id int) (map[string]interface{}, error) {
	var f models.DocumentFlow
	if err := s.DB.Where("is_active = true").First(&f, id).Error; err != nil {
		return nil, err
	}

	// Get associated question groups in order
	var fqgs []models.FlowQuestionGroup
	s.DB.Where("flow_id = ?", f.ID).Order("order_index ASC").Find(&fqgs)

	var groups []map[string]interface{}
	for _, fqg := range fqgs {
		var g models.QuestionGroup
		if err := s.DB.First(&g, fqg.QuestionGroupID).Error; err == nil {
			groups = append(groups, map[string]interface{}{
				"id":            g.ID,
				"name":          g.Name,
				"identifier":    g.Identifier,
				"display_order": g.DisplayOrder,
				"order_index":   fqg.OrderIndex,
			})
		}
	}

	return map[string]interface{}{
		"id":                f.ID,
		"name":              f.Name,
		"description":       f.Description,
		"flow_logic":        f.FlowLogic,
		"starting_group_id": f.StartingGroupID,
		"created_by":        f.CreatedBy,
		"is_active":         f.IsActive,
		"created_at":        f.CreatedAt,
		"updated_at":        f.UpdatedAt,
		"question_groups":   groups,
	}, nil
}

func (s *FlowService) Update(id int, input FlowUpdateInput) (*models.DocumentFlow, error) {
	var f models.DocumentFlow
	if err := s.DB.First(&f, id).Error; err != nil {
		return nil, errors.New("flow not found")
	}

	if input.Name != nil {
		f.Name = *input.Name
	}
	if input.Description != nil {
		f.Description = input.Description
	}
	if input.StartingGroupID != nil {
		f.StartingGroupID = input.StartingGroupID
	}
	if input.FlowLogic != nil {
		f.FlowLogic = input.FlowLogic
	}
	if input.IsActive != nil {
		f.IsActive = *input.IsActive
	}
	f.UpdatedAt = time.Now().UTC()

	if err := s.DB.Save(&f).Error; err != nil {
		return nil, err
	}

	// Update question group associations if provided
	if input.QuestionGroupIDs != nil {
		s.DB.Where("flow_id = ?", f.ID).Delete(&models.FlowQuestionGroup{})
		for i, gid := range input.QuestionGroupIDs {
			fqg := models.FlowQuestionGroup{
				FlowID:          f.ID,
				QuestionGroupID: gid,
				OrderIndex:      i,
			}
			s.DB.Create(&fqg)
		}
	}

	return &f, nil
}

func (s *FlowService) Delete(id int) error {
	var f models.DocumentFlow
	if err := s.DB.First(&f, id).Error; err != nil {
		return errors.New("flow not found")
	}
	f.IsActive = false
	f.UpdatedAt = time.Now().UTC()
	return s.DB.Save(&f).Error
}
